Removing Extra code/comments for outdated lock icon idea and various bug fixes

EverydayBitcoiner opened the pull request:
For list of buttons in the following views added lock icon to the right side of the button if the seed has a passphrase:
-PSBTSelectSeedView
-SeedsMenuView
-ToolsAddressExplorerSelectSourceView

This was a TODO item noted within the code.

Also Fixed bug where back button would not work on SeedSingleSigAddressVerificationSelectSeedView

jdlcdl commented:
This worked for me and looks nice.  

Everywhere that I could think of to "select a loaded seed", I found this new lock icon displayed consistently at the far right edge (the three that you mention at the top of your note plus the one you mention in the last line of your note).

I notice that the lock does NOT appear in the TopNav of the SeedOptionsMenuView when the selected seed has a password.  It doesn't really seem "missing" to me that it's not there, but food for thought if this is a new ui element for seeds w/ a password.

I have yet to form an opinion on whether or not I like the lock icon visible.
At first I like it, it's more info for me.  But, I'll get to know what my seed fingerprints look like without a password, and what they look like with the correct password... so that it will stick out like a sore thumb when they have the wrong password; so I suspect my eyes are on the fingerprint and not the lock.   I'm thinking that I'd prefer the lock NOT being there in the event that Im signing something on mainnet with a friendly over my shoulder... but that's a non-issue since teaching a friend should be done on testnet.

As far as the code changes:
My instinct would be to abstract password like a boolean when treating it like a boolean (if not password:)
Nice catch on the backstack pop, I had not noticed bumping into anything strange there.

Well done eb!

kdmukai commented:
Concept NACK
(despite it being, at some point, my own concept)

I think the comment about adding the lock icon was a bit of an outdated thought.

Those lock icons went from, "Ug, that's hard to do right now" (early in the v0.5.0 dev), to "yeah, no, I don't think they belong in there."

We had discussed the lock icon early with @easyuxd's designs. But as the implementation evolved, I think the idea began to fall out of favor, especially when we made the BIP-39 passphrase a "do it now, or never" option when loading a seed (i.e. there's no routing to go back and set a BIP-39 passphrase if you already "finalized" your seed).

I think it's also likely that this was a holdover concept from the older 4.x original UI where, I believe, a BIP-39 passphrase could be added or removed whenever. In that kind of a UI, a stronger argument can be made for indicating which seeds onboard are currently configured w/a BIP-39 passphrase.

My problem with the lock icon is that it's not really how I think about a BIP-39 passphrase. It suggests that the BIP-39 passphrase is the key that unlocks the account/wallet/etc. Which, in a way, it does. BUT you could, of course, have a number of different BIP-39 passphrases on the same seed. You could push the analogy to say that it IS locked with each of those passphrases, but the thing that's being locked is also inherently changed by each different passphrase. The analogy gets wobbly.

Seed X isn't "locked" by BIP-39 passphrase Y.

Seed X + BIP-39 passphrase Y1 = totally new seed Z1.
Seed X + BIP-39 passphrase Y2 = totally new seed Z2.

I think further cementing the obsolescence of the lock concept is that the UI made a firm decision to identify seeds by their fingerprint. In my mind, that's all the info you need (because, of course, we would have ideally trained people to always note their fingerprints, especially the new fingerprint that results from applying the BIP-39 passphrase).

kdmukai commented:
Another argument: consider a newer user who sets up their funds on a "naked" seed (no BIP-39 passphrase).

They see the lock icon in some tutorials and naturally think, "oh, that must be more secure." So they look to add a BIP-39 passphrase, mistakenly thinking it will add more security for their existing setup.

Or we'll constantly get questions asking, "Hey, how do I lock my seed?" Or even more confusing if asked without context, "Is it okay that my seed is unlocked?"

BIP-39 passphrases are such a huge source of possible problems. I think we have to focus our energy on emphasizing that a BIP-39 passphrase yields a completely different seed. Keeping any relationship with the "naked' seed in the UI makes it just a little bit harder for someone to make the mental leap to realizing it's an entirely new seed now.

EverydayBitcoiner commented:
Sounds good to me. This was just my first little foray into messing with seedsigner and had a bunch of fun messing around.

I reverted my branch and removed the TODO items and code related to adding the lock icon to the seed buttons and kept the bug fix where the back button didn't work when you were on the SeedSingleSigAddressVerificationSelectSeedView.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1499,6 +1499,8 @@ def run(self):
 
         if selected_menu_num == RET_CODE__BACK_BUTTON:
+            # Need to pop extra view from backstack to get to ScanView
+            self.controller.back_stack.pop()
```
A `View` shouldn't assert control over the `Controller`'s `BackStack`.

Probably instead we need to add `skip_current_view=True` to all the `Destination` instances in `AddressVerificationStartView`.

Any `View` that does that kind of auto-routing needs to skip itself as a possible "back" `Destination`.

EverydayBitcoiner replied:
Removed this update and added skip_current_view=True as suggested.

EverydayBitcoiner commented:
@kdmukai, thank you for the feedback. I added `skip_current_view=True` as indicated. That is a more consistent solution.

EverydayBitcoiner commented:
Lol, just noticed that @kdmukai you already fixed the back bug in pull request #262. I'm on a roll here with this pull request haha. Looks like it'll just remove some lines of code and comments in the end.

kdmukai commented:
Oh, wow, @EverydayBitcoiner I didn't even remember that!

I think the Address Explorer has a similar back button issue that afaik is not fixed. Can you take a look there?

EverydayBitcoiner commented:
So I did a little digging and this was the bug that I found for the address explorer (not sure if it is the same one you were thinking of). To start go "Tools">"Address Explorer">"Scan a seed". At this point it will go the the ScanView and then the SeedFinalizeView. When you hit done on the Seed SeedFinalzeView it redirects (through the SeedOptionsView) you back to the SeedExportXpubScriptTypeView for the address explorer path but at the same time it clears the back_stack meaning that the back button takes the user all the way back to home instead of taking the user back to the ToolsAddressExplorerSelectSourceView as would be expected.

I'm not sure what the best fix for that would be. A couple of ideas off the top of my head that may or may not be feasible or desirable:

1. When directing to a side path (such as above) save a copy of what the back stack looks like at that moment and then reset the backstack to that when coming back to the flow.
2. For the SeedFinalizeView figure out a way to only clear the backstack back to the begining of the seed flow instead of completely clearing the backstack. Like maybe whenever a side path is initiated there is a placeholder inserted into the backstack that can then popped back to if needed. This one may be better in that it allows for multiple side paths that will revert back to the most recent placeholder in the backstack.

Thoughts?

This may be a more involved fix rather than an easy one.

Below is the log printout for the views as described above:

```
back_stack: []
Executing MainMenuView()
Appending next destination: ToolsMenuView()
------------------------------
back_stack: [
     0: ToolsMenuView()
]
Executing ToolsMenuView()
Appending next destination: ToolsAddressExplorerSelectSourceView()
------------------------------
back_stack: [
     1: ToolsAddressExplorerSelectSourceView()
     0: ToolsMenuView()
]
Executing ToolsAddressExplorerSelectSourceView()
Appending next destination: ScanView()
------------------------------
back_stack: [
     2: ScanView()
     1: ToolsAddressExplorerSelectSourceView()
     0: ToolsMenuView()
]
Executing ScanView()
render_width: 240
render_height: 240
PiVideoStream: closing everything
Appending next destination: SeedFinalizeView()
------------------------------
back_stack: [
     3: SeedFinalizeView()
     2: ScanView()
     1: ToolsAddressExplorerSelectSourceView()
     0: ToolsMenuView()
]
Executing SeedFinalizeView()
Appending next destination: SeedOptionsView({'seed_num': 0}) | clear_history: True
------------------------------
back_stack: [
     0: SeedOptionsView({'seed_num': 0}) | clear_history: True
]
Executing SeedOptionsView({'seed_num': 0}) | clear_history: True
Skipping current view: SeedOptionsView({'seed_num': 0}) | clear_history: True
Appending next destination: SeedExportXpubScriptTypeView({'seed_num': 0, 'sig_type': 'ss'})
------------------------------
back_stack: [
     0: SeedExportXpubScriptTypeView({'seed_num': 0, 'sig_type': 'ss'})
]
Executing SeedExportXpubScriptTypeView({'seed_num': 0, 'sig_type': 'ss'})
```

jdlcdl commented:
I note that this behavior is the case, but I think I prefer landing back in the main menu after a side-jaunt to load a seed.  The usage of traversing the backstack is inconsistent here and in other places (watch how it grows in Settings) but convenient, at least to me from a user perspective.  There are certainly times when I wouldn't want to go back through steps that have been completed.  Nice work eb, you've hit the ground running!

kdmukai commented:
@EverydayBitcoiner I'll have to take a closer look at the sequence you listed. You may be right that we're over-aggressive in wiping the `BackStack` after loading a seed (however, we definitely need at least *some* clearing; we don't want someone to be able to BACK their way back to the passphrase entry UI, for example).

But in the meantime, the issue in the Address Explorer that I came across was something like: I was working in a different option under Seed Options (playing with a possible Nostr integration), clicked BACK to return to Seed Options, then accidentally selected Address Explorer, clicked BACK to return to Seed Options, but ended up in a weird BACK loop where I had no way to exit Seed Options anymore.

I think the issue was a missing `skip_current_view=True` in the Address Explorer flow, similar to what we discussed above.

EverydayBitcoiner commented:
@kdmukai, I did some more digging it looks like the bug you were referring to was in your new Nostr code and I poked around, fixed it, and made a pull request for you (https://github.com/kdmukai/seedsigner/pull/3). It was just that the back button was adding a view instead of going to the backstack.

As I was poking around looking at that problem I noticed an issue when accessing the Address explorer from a specific seed. It gets some funky behavior because of `resume_main_flow`. When you select a seed and then select "Address Explorer" it sets the `resume_main_flow = Controller.FLOW__ADDRESS_EXPLORER`. If you then press back on the next screen (`SeedExportXpubScriptTypeView`) it does not take you back to the `SeedOptionsView` because the `SeedOptionsView` automatically forwards you to `SeedExportXpubScriptTypeView` if `resume_main_flow = Controller.FLOW__ADDRESS_EXPLORER`. If you press back again it will take you to the `SeedsMenuView` (because when `SeedOptionsView` gets automatically forwarded it doesn't get added to the backstack) and if you select a seed it will again automatically forward you to the `SeedExportXpubScriptTypeView` for address explorer because `resume_main_flow` is still equal to `Controller.FLOW__ADDRESS_EXPLORER`.

The fix I implemented was in the `SeedExportXpubScriptTypeView` where if the back button is pressed it checks to see if the previous view in the backstack is the `SeedsOptionsView` and if it is it will set the `resume_main_flow=None`.  This seems like this should be ok because if a flow is accessed from the Tools section it should `skip_current_view=True` for `SeedOptionsView` so that should only be in the backstack when the flow was accessed from the `SeedOptionsView`. I checked and verified that the Address explorer works from both the tools menu and `SeedOptionsView`

As a side note, it seems that the name of the view `SeedExportXpubScripttypeView` may need to be eventually renamed considering it is now used in more situations than just exporting the Xpub. Maybe something like `SeedScriptTypeSelectView` or something like that.

The other bothersome thing that I encountered was I accidentally scanned an xpub instead of a seed when trying to load a wallet and it gave me no message and just put me at home. I added a message so that if an invalid QR is scanned it gives a message that the scanned qr is invalid or not supported. This seems like a good idea because I could see someone trying to scan a QR and not knowing why it doesn't seem to be working.

EverydayBitcoiner commented:
> I note that this behavior is the case, but I think I prefer that "back" lands me back in the main menu after a side-jaunt to load a seed. The usage of traversing the backstack is inconsistent here and in other places (watch how it grows in Settings) but convenient, at least to me from a user perspective. There are certainly times when I wouldn't want to go back through steps that have been completed. Nice work eb, you've hit the ground running!

I do agree that sometimes I would like to be able to just go to home (especially if I am deep in the settings or going through a flow that I've decided that I want to abort and I've thought that it could be nice to have a back button on the top left and a home button in the top right. I know real estate is limited on that top line and I'm not sure if anyone else has thought about this before.

kdmukai commented:
> [...] I've thought that it could be nice to have a back button on the top left and a home button in the top right. I know real estate is limited on that top line and I'm not sure if anyone else has thought about this before.

There is a tiny shortcut of sorts where if you push LEFT on most screens, the BACK arrow will be highlighted. This is most useful when you're deep in Settings and want to escape!

That being said, a more advanced UI option might be to support long-press. Long-pressing BACK could pop up a dialogue w/an option to return Home.

EverydayBitcoiner commented:
> There is a tiny shortcut of sorts where if you push LEFT on most screens, the BACK arrow will be highlighted. This is most useful when you're deep in Settings and want to escape!

I do use this and it is very nice. Another thought I had was making it so that if you are on the back button and you click left it will go back just as if you clicked the back button. This would make it so you could just click left a bunch of times without having to click/press in between.

> That being said, a more advanced UI option might be to support long-press. Long-pressing BACK could pop up a dialogue w/an option to return Home.

Yes, I've had a couple of times where I thought that long press could be useful for different things. Have you implemented a long click anywhere?

kdmukai commented:
> Another thought I had was making it so that if you are on the back button and you click left it will go back just as if you clicked the back button.

For the current hardware, I think the joystick is still a little iffy and too easy to accidentally push a direction that wasn't intended. So for now I'd still want the BACK process to be more explicit/intentional.

> Have you implemented a long click anywhere?

No, not yet. It would have to be integrated into the timings for `HardwareButtons` with some complex-ish interplay with the `check_release` option. But should be do-able. I'd want @easyuxd's thoughts on it first, though. @EverydayBitcoiner, maybe move this to its own github Issue and we can discuss further there?

EverydayBitcoiner commented:
> No, not yet. It would have to be integrated into the timings for HardwareButtons with some complex-ish interplay with the check_release option. But should be do-able. I'd want @easyuxd's thoughts on it first, though. @EverydayBitcoiner, maybe move this to its own github Issue and we can discuss further there?

Opened up Issue https://github.com/SeedSigner/seedsigner/issues/287 for long click functionality.

EverydayBitcoiner commented:
I also made a change on the ScanView to match other qr scan screens (see below):

![image](https://user-images.githubusercontent.com/119913286/209714315-1a132bf5-9eaa-40bf-aa99-353bee0c5b63.png).

@kdmukai, what would you like to see on this draft PR before I push it to review again?

EverydayBitcoiner commented on src/seedsigner/gui/screens/scan_screens.py:
```diff
@@ -18,7 +18,7 @@
 @dataclass
 class ScanScreen(BaseScreen):
     decoder: DecodeQR = None
-    instructions_text: str = "Scan a QR code"
```
Issue #185

EverydayBitcoiner commented on src/seedsigner/gui/screens/scan_screens.py:
```diff
@@ -18,7 +18,7 @@
 class ScanScreen(BaseScreen):
     decoder: DecodeQR = None
-    instructions_text: str = "Scan a QR code"
+    instructions_text: str = "< back | Scan a QR code"
```
Update to match seed image capture screen. Issue #185

jdlcdl replied:
TY @EverydayBitcoiner, this improvement resolving an old issue has been borrowed in pr #323

EverydayBitcoiner commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -41,9 +41,6 @@ def run(self):
             
-            if seed.passphrase is not None:
-                # TODO: Include lock icon on right side of button
-                pass
```
Decided that it was not desired to have a lock icon as described in this comment (see [here](https://github.com/SeedSigner/seedsigner/pull/286#issuecomment-1363351268) for discussion). Removing comment and if statment.

jdlcdl replied:
TY @EverydayBitcoiner,  this improvement has been borrowed (as an elif at same level) in pr #323

jdlcdl replied:
TY @EverydayBitcoiner, this code cleanup has been included in pr #324

EverydayBitcoiner commented on src/seedsigner/views/scan_views.py:
```diff
@@ -102,6 +102,9 @@ def run(self):
 
+        if self.decoder.is_invalid:
+            raise Exception("QRCode not recognized or not yet supported.")
+
```
Added better note for when an unrecognized QR is scanned. Currently is simply goes to home without any message.

jdlcdl replied:
I'm a fan of this.

I frequently run into this.  In every case so far it's been user error... but I wasn't sure if it might be some new bug I've introduced while testing elsewhere, or if I'm indeed scanning the wrong qrcode (most recently, trying to load a single seed's xpub in Sparrow instead of the wallet descriptor xpub just above).   I cannot be the only one that makes this mistake frequently and suspects "bug in seedsigner" when "bug in my head" is the better diagnosis.  This exception will lead me at least, to question what I'm doing instead of retrying the same mistake and hoping for better results.

EverydayBitcoiner replied:
💯

Always nice to have clearer messages.

jdlcdl replied:
TY @EverydayBitcoiner, this improvement has been borrowed (as an elif at same level) in pr #323

EverydayBitcoiner commented on src/seedsigner/views/seed_views.py:
```diff
@@ -37,7 +37,6 @@ def __init__(self):
         for seed in self.controller.storage.seeds:
             self.seeds.append({
                 "fingerprint": seed.get_fingerprint(self.settings.get_value(SettingsConstants.SETTING__NETWORK)),
-                "has_passphrase": seed.passphrase is not None
```
Not needed since no longer wanting to add lock icon for seeds with passphrases (see [here](https://github.com/SeedSigner/seedsigner/pull/286#issuecomment-1363351268).Removing unnecessary code.

jdlcdl replied:
TY @EverydayBitcoiner, this code cleanup, less the trailing comma on the previous item, has been included in pr #324

EverydayBitcoiner commented on src/seedsigner/views/seed_views.py:
```diff
@@ -536,6 +535,9 @@ def run(self):
         if selected_menu_num == RET_CODE__BACK_BUTTON:
+            # If previous view is SeedOptionsView then that should be where resume_main_flow started (otherwise it would have been skipped).
+            if self.controller.back_stack[-2].View_cls == SeedOptionsView:
+                self.controller.resume_main_flow = None
```
Clears "resume_main_flow" variable when back button goes to "SeedOptionsView" (see [here](https://github.com/SeedSigner/seedsigner/pull/286#issuecomment-1364552409) for full explanation.

jdlcdl replied:
TY @EverydayBitcoiner.  I'm grabbing this snippet of code (and adding a check to avoid an IndexError) for a separate/smaller pr that fixes having to back-out-twice from AddressExplorer when coming from SeedOptionsView menu.

EverydayBitcoiner commented on src/seedsigner/views/tools_views.py:
```diff
@@ -434,9 +434,6 @@ def run(self):
             
-            if seed.passphrase is not None:
-                # TODO: Include lock icon on right side of button
-                pass
```
Decided that it was not desired to have a lock icon as described in this comment (see [here](https://github.com/SeedSigner/seedsigner/pull/286#issuecomment-1363351268) for discussion). Removing comment and if statment.

jdlcdl replied:
TY @EverydayBitcoiner, this code cleanup has been included in pr #324

EverydayBitcoiner commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1484,9 +1486,6 @@ def run(self):
             
-            if seed.passphrase is not None:
-                # TODO: Include lock icon on right side of button
-                pass
```
Decided that it was not desired to have a lock icon as described in this comment (see [here](https://github.com/SeedSigner/seedsigner/pull/286#issuecomment-1363351268) for discussion). Removing comment and if statment.

jdlcdl replied:
TY @EverydayBitcoiner, this code cleanup has been included in pr #324

EverydayBitcoiner commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1398,22 +1400,22 @@ def run(self):
         elif self.controller.unverified_address["script_type"] == SettingsConstants.LEGACY_P2PKH:
             # TODO: detect single sig vs multisig or have to prompt?
-            return Destination(NotYetImplementedView)
+            return Destination(NotYetImplementedView, skip_current_view=True)
```
This View should not be added to the backstack (see [here](https://github.com/SeedSigner/seedsigner/pull/286#discussion_r1056014645) for discussion).

jdlcdl replied:
TY @EverydayBitcoiner, The top 3 changes to skip_current_view=True from this View will be borrowed and included in a smaller pr.  I'll edit this later.  In all 3 cases, these additions allow the user to exit to the main menu via the back button, instead of landing in the scan view for the first two, or landing in the same screen forever in the last one.

Note for adding ", skip_current_view=True" to Destination(NotYetImplementedView) is currently NEVER required because regardless of how the user exits that screen, via "ok" or "back" the next step is directly to the MainMenuView.

jdlcdl commented:
Hey @EverydayBitcoiner,

I know that this started out as a much more involved pull request, but after reviewing changes you've made, the end result seems like good code cleanup with worth-while improvements.  Can you correct me if I'm misreading the changes as they stand currently (as of you marking this pull request ready for review again last week)?

* When clicking "Scan" from the main menu, this includes a hint that "back" is a way to escape?  If so, I like this.

* code clean-up in PSBTSelectSeedView and in SeedSingleSigAddressVerificationSelectSeedView for obsolete 'future todo' hooks -- originating from when the lock icon was planned to be visible for seeds with a passphrase set.  If so, I like these.

* When having scanned an invalid QR Code, a visible exception is raised instead of landing back in the Main Menu.  Call me insane, but whenever this happens to me, I keep trying a few more times -- expecting different results.  I really DO like the idea of having a visible error screen for an invalid QR, so that I can immediately figure out that I've chosen the wrong QR Code to scan.  I've done this a few times with Sparrow, scanning the xpub instead of the multisig wallet descriptor.

* In the SeedsMenuView, it has code cleanup of an unnecessary dictionary key which is obsoleted now that a lock icon is no longer wanted next to seeds with a passphrase set.  If so, I like this.

* **Im still puzzled on the controller.resume_main_flow behavior change**, even though I read your explanation.  I'd like to be able to anticipate exactly what it's doing so that I could test this.

* And from the AddressVerification flow, changes that would **avoid landing the user on the "scan" screen as they back-out**???

If you're around, please confirm that I understand these changes so that I can get to testing them asap.  You can also ping me in the telegram new devs group.  See you soon!

jdlcdl commented:
4 pull requests since yesterday, pr #322, pr #323, pr #324 and pr #325 were all derived by picking parts of this pull-request.  If they are all merged, or if they are closed without being merged, then this one can also be closed.

Thank you @EverydayBitcoiner for all of your hard work here, implemented during the holidays nonetheless; it is much appreciated!

newtonick commented:
The work in this PR was done in other smaller PR's. Credit to @jdlcdl  and @EverydayBitcoiner for this work.
